### PR TITLE
Integrate FFS changes on savedsites for sync

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/488551667048375/1205741811092048/f

### Description
Implements FFS logic on sync.

Reading the guide [here](https://app.asana.com/0/0/1205552640876411/f) should give you a good overview on all the changes introduced.

### Steps to test this PR
(Download latest build from macos https://app.asana.com/0/0/1205470563151378/f)

_Feature 1_
- [x] Using the same account on macOS and on your phone
- [x] when using not-unified mode
- [x] you should see favorites created on your device
- [x] reordering them will not affect other clients
- [x] try adding a few in both platforms and ensure you don't receive them in the other client
- [x] Now using unified mode
- [x] both devices should see the same bookmarks
- [x] reordering favorites will be reflected in both devices
- [x] try adding/remove/moving favorites and ensure the other device receives the changes
- [x] In this mode, any favorite you add will also be a "native" favorite
- [x] changing to not-unified mode, you will also see the favorites you added during unified view on that client

[Enabling sync](https://app.asana.com/0/0/1205419026232006/f)
- [x] Ensure when creating account or joining account
- [x] device should never send foreign favorites (in our case is desktop)
- [x] Any interaction from the user should never include a change on that folder 

_Inserting/deleting favorites_
- [x] check the diagram in this [task](https://app.asana.com/0/0/1205518566847038/f)
- [x] While in unified mode
- [x] adding a favorite will add it to root and mobile
- [x] when removing the favorite, will remove it from any favorite folder
- [x] While in native mode
- [x] adding a favorite will add it to root and mobile
- [x] if the favorite is a desktop fav, it's also a bookmark, you can visit the bookmark and ensure is not a mobile favorite
- [x] Adding it to favorites, will add the favorite to native folder
- [x] if the favorite is present in all 3 folders, removing it from mobile in native mode, should only remove it from mobile folder, not from root or desktop

_Disabling Sync_
- [x] [here's the task with the specs](https://app.asana.com/0/0/1205470518820708/f)
- [x] If using unified mode
- [x] when disabling sync, your favorites should be the same as before (aka. favorites_root)
- [x] if using native mode
- [x] favorites_root will be all your mobile_favorite_root favorites

### UI changes
| Before  | After |
| ------ | ----- |
!(Upload before screenshot)|(Upload after screenshot)|

### CODE REVIEW

**Comment by cmonfortep** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3704** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3704" target="_blank">Graphite</a> 
    * **PR #3705** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3705" target="_blank">Graphite</a> 
      * **PR #3707** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3707" target="_blank">Graphite</a> 
        * **PR #3710** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3710" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3710?utm_source=stack-comment).

**Review by malmstein — CHANGES_REQUESTED** [review]
As discussed offline:

- Rename SavedSitesSettingsRepository  to FavoritesFormFactorSettingsRepository
- Rename FavoritesAccessor to FavoritesDelegate 
- Move ViewMode un FavoritesAccessorto SavedSitesSettingsRepository
- Move ffs setting logic from FavoritesAccessor to SavedSitesSettingsRepository
-   - getInsertFavoriteFolder
-   - and all other functions that return a folder (String, List<String>)

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/FavoritesAccessor.kt [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+import com.squareup.anvil.annotations.*
+import io.reactivex.*
+import java.util.*
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
+import timber.log.*
+
+interface FavoritesAccessor {
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/FavoritesAccessor.kt]: Skipping this one, as this is an internal class, and just a delegate. Docs are reflected in SavedSitesRepo api

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/SyncEntitiesExtension.kt
Comment (malmstein) [inline saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/SyncEntitiesExtension.kt:1]: this is not a SyncEntity, but an Entity. I’d rename this file to `EntitiesExtensions`

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/SyncSavedSitesRepository.kt
Comment (malmstein) [inline saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/SyncSavedSitesRepository.kt:1]: Why do we need this? what’s the difference with the current Repository?

**Review by malmstein — APPROVED** [review]
Looks good! I’m not sure about the `SyncSavedSitesRepository`, but it’s nothing that blocks this.

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/SyncSavedSitesRepository.kt
Comment (cmonfortep) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/SyncSavedSitesRepository.kt:1]: As discussed, this repository will replace the current repository in sync related classes (aka. provider and persisters). Current repository will only be used by saved-sites feature (non-sync)

**Pull request merged**